File: packages/judgment-core/src/policy-engine.ts
import { moreRestrictive } from './materiality.js';
import type {
  AuthorityMode,
  DelegationConditions,
  InterventionLevel,
  JudgmentPoint,
  JudgmentPolicy,
  PolicyRule,
  PolicyScope,
  RuleCondition,
} from './types.js';

// ---------------------------------------------------------------------------
// Result types
// ---------------------------------------------------------------------------

export interface PolicyRuleMatch {
  readonly policy: JudgmentPolicy;
  readonly rule: PolicyRule;
}

export interface PolicyEvalResult {
  readonly intervention: InterventionLevel;
  readonly authority?: AuthorityMode;
  readonly delegationConditions?: DelegationConditions;
  readonly matchedRules: readonly PolicyRuleMatch[];
}

// ---------------------------------------------------------------------------
// Scope matching
// ---------------------------------------------------------------------------

/**
 * Check whether a judgment point falls within a policy's scope.
 * All specified scope conditions must match (AND logic).
 * Omitted conditions match everything.
 */
export function matchesScope(point: JudgmentPoint, scope: PolicyScope): boolean {
  if (
    scope.categories != null &&
    scope.categories.length > 0 &&
    !scope.categories.includes(point.category)
  ) {
    return false;
  }

  if (
    scope.triggerSources != null &&
    scope.triggerSources.length > 0 &&
    !scope.triggerSources.includes(point.trigger.source)
  ) {
    return false;
  }

  if (scope.artifactTypes != null && scope.artifactTypes.length > 0) {
    // Check if any of the point's evidence refs or affected artifacts match.
    // Since the point schema tracks artifact IDs, not types, we check evidenceRefs.
    const pointArtifactTypes = new Set((point.evidenceRefs ?? []).map((ref) => ref.artifactType));
    const hasMatch = scope.artifactTypes.some((t) => pointArtifactTypes.has(t));
    if (!hasMatch) {
      return false;
    }
  }

  if (scope.materialityScoreMin != null && point.materiality.score < scope.materialityScoreMin) {
    return false;
  }

  if (scope.materialityScoreMax != null && point.materiality.score > scope.materialityScoreMax) {
    return false;
  }

  return true;
}

// ---------------------------------------------------------------------------
// Rule condition matching
// ---------------------------------------------------------------------------

/**
 * Evaluate a rule condition against a judgment point.
 * All specified fields must match (AND logic).
 * Omitted fields match everything.
 */
export function evaluateRuleCondition(point: JudgmentPoint, condition: RuleCondition): boolean {
  if (
    condition.materialityScoreMin != null &&
    point.materiality.score < condition.materialityScoreMin
  ) {
    return false;
  }

  if (
    condition.materialityScoreMax != null &&
    point.materiality.score > condition.materialityScoreMax
  ) {
    return false;
  }

  if (condition.dimensionThresholds != null) {
    const dims = point.materiality.dimensions;
    const thresholds = condition.dimensionThresholds;
    // "If any specified dimension meets or exceeds its threshold, the condition matches."
    let anyMeetsThreshold = false;
    for (const [key, threshold] of Object.entries(thresholds)) {
      if (threshold != null) {
        const dimKey = key as keyof typeof dims;
        if (dims[dimKey] >= threshold) {
          anyMeetsThreshold = true;
          break;
        }
      }
    }
    // If thresholds were specified but none met, condition fails.
    const hasThresholds = Object.values(thresholds).some((v) => v != null);
    if (hasThresholds && !anyMeetsThreshold) {
      return false;
    }
  }

  if (condition.hardTrigger != null) {
    if (
      point.trigger.hardTrigger !== condition.hardTrigger &&
      point.materiality.hardTrigger !== condition.hardTrigger
    ) {
      return false;
    }
  }

  if (
    condition.categories != null &&
    condition.categories.length > 0 &&
    !condition.categories.includes(point.category)
  ) {
    return false;
  }

  // expression is not evaluated in this implementation (spec says implementations may define their own DSL)

  return true;
}

// ---------------------------------------------------------------------------
// Single policy evaluation
// ---------------------------------------------------------------------------

/**
 * Evaluate a single policy against a judgment point.
 * Returns the first matching rule, or null if no rule matches.
 */
export function evaluatePolicy(point: JudgmentPoint, policy: JudgmentPolicy): PolicyRule | null {
  if (!policy.enabled) {
    return null;
  }

  if (!matchesScope(point, policy.scope)) {
    return null;
  }

  for (const rule of policy.rules) {
    if (evaluateRuleCondition(point, rule.condition)) {
      return rule;
    }
  }

  return null;
}

// ---------------------------------------------------------------------------
// Multi-policy evaluation
// ---------------------------------------------------------------------------

const AUTHORITY_RANK: Record<AuthorityMode, number> = {
  rule: 0,
  delegated: 1,
  collaborative: 2,
  human: 3,
};

function moreRestrictiveAuthority(a: AuthorityMode, b: AuthorityMode): AuthorityMode {
  return AUTHORITY_RANK[a] >= AUTHORITY_RANK[b] ? a : b;
}

/**
 * Evaluate all applicable policies for a judgment point and resolve conflicts.
 *
 * Policies are sorted by priority (lower = higher precedence).
 * At each precedence level, the most restrictive intervention and authority win.
 * Delegation is disallowed if any matching policy disallows it.
 */
export function evaluatePolicies(
  point: JudgmentPoint,
  policies: readonly JudgmentPolicy[],
): PolicyEvalResult {
  // Gather all matching rules.
  const matches: PolicyRuleMatch[] = [];

  for (const policy of policies) {
    const rule = evaluatePolicy(point, policy);
    if (rule != null) {
      matches.push({ policy, rule });
    }
  }

  if (matches.length === 0) {
    // Fall back to score-based defaults (caller is responsible for applying default thresholds).
    return {
      intervention: point.materiality.interventionLevel ?? 'trace',
      matchedRules: [],
    };
  }

  return resolveConflicts(matches);
}

/**
 * Resolve conflicts among matching policy rules.
 *
 * - Sorted by policy priority (lower = higher precedence).
 * - At equal precedence, most restrictive intervention and authority win.
 * - Any delegation disallow overrides all.
 */
export function resolveConflicts(matches: readonly PolicyRuleMatch[]): PolicyEvalResult {
  // Sort by priority (ascending = higher precedence first).
  const sorted = [...matches].sort((a, b) => a.policy.priority - b.policy.priority);

  const first = sorted[0];
  if (first == null) {
    return { intervention: 'trace', matchedRules: [] };
  }

  let intervention: InterventionLevel = first.rule.intervention;
  let authority: AuthorityMode | undefined = first.rule.authorityOverride?.mode;
  let delegationConditions: DelegationConditions | undefined = first.rule.delegationConditions;
  let delegationAllowed = delegationConditions?.allowed ?? true;

  for (let i = 1; i < sorted.length; i++) {
    const match = sorted[i];
    if (match == null) continue;
    const rule = match.rule;

    // Same or lower priority: apply most-restrictive-wins.
    intervention = moreRestrictive(intervention, rule.intervention);

    if (rule.authorityOverride?.mode != null) {
      authority =
        authority != null
          ? moreRestrictiveAuthority(authority, rule.authorityOverride.mode)
          : rule.authorityOverride.mode;
    }

    if (rule.delegationConditions != null) {
      if (rule.delegationConditions.allowed === false) {
        delegationAllowed = false;
      }
      // Merge: most restrictive maxMaterialityScore, highest requiredConfidence, union of excludedCategories.
      delegationConditions = mergeDelegationConditions(
        delegationConditions,
        rule.delegationConditions,
      );
    }
  }

  if (delegationConditions != null && !delegationAllowed) {
    delegationConditions = { ...delegationConditions, allowed: false };
  }

  return {
    intervention,
    authority,
    delegationConditions,
    matchedRules: sorted,
  };
}

function mergeDelegationConditions(
  a: DelegationConditions | undefined,
  b: DelegationConditions,
): DelegationConditions {
  if (a == null) return b;

  const maxScore =
    a.maxMaterialityScore != null && b.maxMaterialityScore != null
      ? Math.min(a.maxMaterialityScore, b.maxMaterialityScore)
      : (a.maxMaterialityScore ?? b.maxMaterialityScore);

  const reqConf =
    a.requiredConfidence != null && b.requiredConfidence != null
      ? Math.max(a.requiredConfidence, b.requiredConfidence)
      : (a.requiredConfidence ?? b.requiredConfidence);

  const excludedSet = new Set([...(a.excludedCategories ?? []), ...(b.excludedCategories ?? [])]);

  return {
    allowed: (a.allowed ?? true) && (b.allowed ?? true),
    maxMaterialityScore: maxScore,
    requiredConfidence: reqConf,
    excludedCategories: excludedSet.size > 0 ? Array.from(excludedSet) : undefined,
    requiresPriorHumanResolution:
      (a.requiresPriorHumanResolution ?? false) || (b.requiresPriorHumanResolution ?? false),
    auditRequired: (a.auditRequired ?? true) || (b.auditRequired ?? true),
  };
}

File: backend/judgment_core/src/judgment_core/policy_engine.py
"""Policy engine for evaluating judgment policies against judgment points.

Policies are sorted by priority (lower = higher precedence).
At each precedence level, the most restrictive intervention and authority win.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from judgment_core.materiality import more_restrictive
from judgment_core.types import (
    AuthorityMode,
    DelegationConditions,
    InterventionLevel,
    JudgmentPoint,
    JudgmentPolicy,
    PolicyRule,
    PolicyScope,
    RuleCondition,
)

# ---------------------------------------------------------------------------
# Result types
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class PolicyRuleMatch:
    policy: JudgmentPolicy
    rule: PolicyRule


@dataclass(frozen=True)
class PolicyEvalResult:
    intervention: InterventionLevel
    matched_rules: list[PolicyRuleMatch] = field(default_factory=list)
    authority: AuthorityMode | None = None
    delegation_conditions: DelegationConditions | None = None


# ---------------------------------------------------------------------------
# Scope matching
# ---------------------------------------------------------------------------


def matches_scope(point: JudgmentPoint, scope: PolicyScope) -> bool:
    """Check whether a judgment point falls within a policy's scope.

    All specified scope conditions must match (AND logic).
    Omitted conditions match everything.
    """
    if scope.categories and point.category.value not in [c.value for c in scope.categories]:
        return False

    if scope.trigger_sources and point.trigger.source.value not in [
        s.value for s in scope.trigger_sources
    ]:
        return False

    if scope.artifact_types:
        point_artifact_types = {ref.artifact_type.value for ref in (point.evidence_refs or [])}
        if not any(t.value in point_artifact_types for t in scope.artifact_types):
            return False

    if (
        scope.materiality_score_min is not None
        and point.materiality.score < scope.materiality_score_min
    ):
        return False

    return not (
        scope.materiality_score_max is not None
        and point.materiality.score > scope.materiality_score_max
    )


# ---------------------------------------------------------------------------
# Rule condition matching
# ---------------------------------------------------------------------------


def evaluate_rule_condition(point: JudgmentPoint, condition: RuleCondition) -> bool:
    """Evaluate a rule condition against a judgment point.

    All specified fields must match (AND logic).
    Omitted fields match everything.
    """
    if (
        condition.materiality_score_min is not None
        and point.materiality.score < condition.materiality_score_min
    ):
        return False

    if (
        condition.materiality_score_max is not None
        and point.materiality.score > condition.materiality_score_max
    ):
        return False

    if condition.dimension_thresholds is not None:
        dims = point.materiality.dimensions
        thresholds = condition.dimension_thresholds
        threshold_map = {
            "methodological_discretion": thresholds.methodological_discretion,
            "downstream_influence": thresholds.downstream_influence,
            "uncertainty": thresholds.uncertainty,
            "consequence": thresholds.consequence,
            "reversibility": thresholds.reversibility,
            "accountability_requirement": thresholds.accountability_requirement,
        }
        any_meets = False
        has_thresholds = False
        for key, threshold in threshold_map.items():
            if threshold is not None:
                has_thresholds = True
                if getattr(dims, key) >= threshold:
                    any_meets = True
                    break
        if has_thresholds and not any_meets:
            return False

    if condition.hard_trigger is not None and (
        point.trigger.hard_trigger != condition.hard_trigger
        and point.materiality.hard_trigger != condition.hard_trigger
    ):
        return False

    return not (
        condition.categories and point.category.value not in [c.value for c in condition.categories]
    )


# ---------------------------------------------------------------------------
# Single policy evaluation
# ---------------------------------------------------------------------------


def evaluate_policy(point: JudgmentPoint, policy: JudgmentPolicy) -> PolicyRule | None:
    """Evaluate a single policy against a judgment point.

    Returns the first matching rule, or None if no rule matches.
    """
    if not policy.enabled:
        return None

    if not matches_scope(point, policy.scope):
        return None

    for rule in policy.rules:
        if evaluate_rule_condition(point, rule.condition):
            return rule

    return None


# ---------------------------------------------------------------------------
# Multi-policy evaluation
# ---------------------------------------------------------------------------

_AUTHORITY_RANK: dict[AuthorityMode, int] = {
    AuthorityMode.rule: 0,
    AuthorityMode.delegated: 1,
    AuthorityMode.collaborative: 2,
    AuthorityMode.human: 3,
}


def _more_restrictive_authority(a: AuthorityMode, b: AuthorityMode) -> AuthorityMode:
    return a if _AUTHORITY_RANK[a] >= _AUTHORITY_RANK[b] else b


def evaluate_policies(
    point: JudgmentPoint,
    policies: list[JudgmentPolicy],
) -> PolicyEvalResult:
    """Evaluate all applicable policies for a judgment point and resolve conflicts.

    Policies are sorted by priority (lower = higher precedence).
    At each precedence level, the most restrictive intervention and authority win.
    Delegation is disallowed if any matching policy disallows it.
    """
    matches: list[PolicyRuleMatch] = []

    for policy in policies:
        rule = evaluate_policy(point, policy)
        if rule is not None:
            matches.append(PolicyRuleMatch(policy=policy, rule=rule))

    if not matches:
        return PolicyEvalResult(
            intervention=point.materiality.intervention_level or InterventionLevel.trace,
            matched_rules=[],
        )

    return resolve_conflicts(matches)


def resolve_conflicts(matches: list[PolicyRuleMatch]) -> PolicyEvalResult:
    """Resolve conflicts among matching policy rules.

    - Sorted by policy priority (ascending = higher precedence first).
    - At equal precedence, most restrictive intervention and authority win.
    - Any delegation disallow overrides all.
    """
    sorted_matches = sorted(matches, key=lambda m: m.policy.priority)

    if not sorted_matches:
        return PolicyEvalResult(intervention=InterventionLevel.trace, matched_rules=[])

    first = sorted_matches[0]
    intervention = InterventionLevel(first.rule.intervention.value)
    authority: AuthorityMode | None = (
        AuthorityMode(first.rule.authority_override.mode.value)
        if first.rule.authority_override
        else None
    )
    delegation_conditions = first.rule.delegation_conditions
    delegation_allowed = delegation_conditions.allowed if delegation_conditions else True

    for match in sorted_matches[1:]:
        rule = match.rule
        rule_intervention = InterventionLevel(rule.intervention.value)
        intervention = more_restrictive(intervention, rule_intervention)

        if rule.authority_override and rule.authority_override.mode:
            rule_authority = AuthorityMode(rule.authority_override.mode.value)
            authority = (
                _more_restrictive_authority(authority, rule_authority)
                if authority is not None
                else rule_authority
            )

        if rule.delegation_conditions is not None:
            if rule.delegation_conditions.allowed is False:
                delegation_allowed = False
            delegation_conditions = _merge_delegation_conditions(
                delegation_conditions, rule.delegation_conditions
            )

    if delegation_conditions is not None and not delegation_allowed:
        delegation_conditions = delegation_conditions.model_copy(update={"allowed": False})

    return PolicyEvalResult(
        intervention=intervention,
        authority=authority,
        delegation_conditions=delegation_conditions,
        matched_rules=sorted_matches,
    )


def _merge_delegation_conditions(
    a: DelegationConditions | None,
    b: DelegationConditions,
) -> DelegationConditions:
    if a is None:
        return b

    max_score: float | None
    if a.max_materiality_score is not None and b.max_materiality_score is not None:
        max_score = min(a.max_materiality_score, b.max_materiality_score)
    else:
        max_score = (
            a.max_materiality_score
            if a.max_materiality_score is not None
            else b.max_materiality_score
        )

    req_conf: float | None
    if a.required_confidence is not None and b.required_confidence is not None:
        req_conf = max(a.required_confidence, b.required_confidence)
    else:
        req_conf = (
            a.required_confidence if a.required_confidence is not None else b.required_confidence
        )

    excluded = set((a.excluded_categories or []) + (b.excluded_categories or []))

    return DelegationConditions(
        allowed=(a.allowed if a.allowed is not None else True)
        and (b.allowed if b.allowed is not None else True),
        max_materiality_score=max_score,
        required_confidence=req_conf,
        excluded_categories=list(excluded) if excluded else None,
        requires_prior_human_resolution=(a.requires_prior_human_resolution or False)
        or (b.requires_prior_human_resolution or False),
        audit_required=(a.audit_required if a.audit_required is not None else True)
        or (b.audit_required if b.audit_required is not None else True),
    )
